projects/adventure/adv.py

```diff
@@ -17,20 +17,114 @@
 world.loadGraph(roomGraph)
 
 # UNCOMMENT TO VIEW MAP
-world.printRooms()
+# world.printRooms()
 
 player = Player("Name", world.startingRoom)
 
+class Queue():
+    def __init__(self):
+        self.queue = []
+    def enqueue(self, value):
+        self.queue.append(value)
+    def dequeue(self):
+        if self.size() > 0:
+            return self.queue.pop(0)
+        else:
+            return None
+    def size(self):
+        return len(self.queue)
+
+
 # Fill this out
 traversalPath = []
 
+graph = {}
+
+def backtrack(direction):
+    if direction == "n":
+        return "s"
+    elif direction == "s":
+        return "n"
+    elif direction == "w":
+        return "e"
+    elif direction == "e":
+        return "w"
+
+
+def bft(graph, starting_room):
+    q = Queue()
+    # Create an empty visited dictionary 
+    visited = set() 
+    # Add a path to the starting vertex of the queue
+    path_traverse = []
+    q.enqueue([starting_room])
+    # While the queue is not empty
+    while q.size() > 0:
+        # Dequeue first path
+        path = q.dequeue()
+        # Get last vertex
+        cur_room = path[-1]
+        if cur_room not in visited:
+            visited.add(cur_room)
+            for room in graph[cur_room]:
+                if graph[cur_room][room] == "?":
+                    return path
+            for room_exit in graph[cur_room]:
+                path_traverse.append(room_exit)
+                next_room = graph[cur_room][room_exit]
+                path_copy = path.copy()
+                path_copy.append(next_room)
+                q.enqueue(path_copy)
+
+
+# While graph is fewer than 500 rooms
+while len(graph) < len(roomGraph):
+    # Save the current room ID
+    currentRoomID = player.currentRoom.id
+    # If the current room is not in the graph yet
+    if currentRoomID not in graph:
+        # Put the room into the graph with no exits yet
+        graph[currentRoomID] = {}
+        # For each exit in the room's available exits
+        for exit in player.currentRoom.getExits():
+            # Set all exit values to '?' for the first time visiting the room
+            graph[currentRoomID][exit] = "?"
```

Comment:
Consider the case where currentRoom is not in the visited set, but one of the rooms it has an exit to is.  Do we really want to mark this already visited exit as a '?', or is there a way we can check that and handle it appropriately?